jps-plugin/src/org/intellij/erlang/jps/builder/ErlangModuleBuildOrderBuilder.java

```diff
@@ -0,0 +1,199 @@
+    }
+
+
+    if (dirtyErlangFilePaths.isEmpty()) {
+      LOG.debug("There aren't dirty .erl or .hrl files.");
+    }
+    else {
+      LOG.debug("Search dirty modules.");
+      List<String> sortedDirtyModules = getSortedDirtyModules(context, projectBuildOrder, dirtyErlangFilePaths);
+      addFilesToBuiltTarget(context, sortedDirtyModules);
+    }
+
+  }
+
+  @NotNull
+  @Override
+  public String getPresentableName() {
+    return NAME;
+  }
+
+  @Nullable
+  private static ErlangProjectBuildOrder loadProjectBuildOrder(@NotNull CompileContext context) {
+    return readFromXML(context, BUILD_ORDER_FILE_NAME, ErlangProjectBuildOrder.class);
+  }
+
+  private static boolean isSourceOrHeader(@NotNull File file) {
+    return file.getName().endsWith(".erl") || file.getName().endsWith(".hrl");
+  }
+
+  private static List<String> getSortedDirtyModules(@NotNull CompileContext context,
+                                                    @NotNull ErlangProjectBuildOrder projectBuildOrder,
+                                                    @NotNull List<String> dirtyErlangFilePaths) {
+    SortedModuleDependencyGraph graph = new SortedModuleDependencyGraph(projectBuildOrder);
+    Set<String> allDirtyFiles = getAllDirtyFiles(dirtyErlangFilePaths, GraphGenerator.create(graph));
+    return getSortedDirtyModules(projectBuildOrder.myErlangFiles, allDirtyFiles);
+  }
+
+
+  private static void addPrepareDependenciesFailedMessage(@NotNull CompileContext context) {
+    context.processMessage(new CompilerMessage(NAME, BuildMessage.Kind.WARNING, "The project will be fully rebuilt due to errors."));
+  }
+
+  private static Set<String> getAllDirtyFiles(@NotNull List<String> dirtyFiles,
+                                              @NotNull GraphGenerator<String> graph) {
+    Set<String> allDirtyFiles = ContainerUtil.newHashSet();
+    for (String node : dirtyFiles) {
+      if (node != null) {
+        findDirtyFiles(node, graph, allDirtyFiles);
+      }
+    }
+    return allDirtyFiles;
+  }
+
+  private static void findDirtyFiles(@NotNull String filePath,
+                                     @NotNull GraphGenerator<String> graph,
+                                     @NotNull Set<String> dirtyFiles) {
+    if (dirtyFiles.contains(filePath)) return;
+    dirtyFiles.add(filePath);
+    Iterator<String> childIterator = graph.getOut(filePath);
+    while (childIterator.hasNext()) {
+      findDirtyFiles(childIterator.next(), graph, dirtyFiles);
+    }
+  }
+
+  @NotNull
+  private static List<String> getSortedDirtyModules(@NotNull List<ErlangFileDescriptor> sortedFiles,
+                                                    @NotNull final Set<String> allDirtyFiles) {
+    return ContainerUtil.mapNotNull(sortedFiles, new Function<ErlangFileDescriptor, String>() {
+      @Nullable
+      @Override
+      public String fun(ErlangFileDescriptor node) {
+        return node.myPath.endsWith(".erl") && allDirtyFiles.contains(node.myPath) ? node.myPath : null;
```

Comment:
Please, reuse your extracted method for checking if a file is an Erlang source file.